From f14e02251229a7afa6d2c5bd3fc3cf9301a6a3aa Mon Sep 17 00:00:00 2001
From: rleix <rayx.lei@intel.com>
Date: Thu, 11 Jan 2018 14:13:57 +0800
Subject: [PATCH 2/2] Forcibly collapse the statusbar when incoming call
 received.

Avoiding the fullscreen statusbar hide the incoming call.
Simply call collapsePanels to collapse the statusbar. Then the
incoming call will be displayed.

Test:
1. Pair reference phone to DUT via BT.
2. Tap the overview icon(to be in home screen).
3. Receive a call from another phone to reference phone(Which is
paired to DUT) and check whether the incoming call ui displayed.

Change-Id: Iefde8deb6ac91fae753f6f1b7f589d3a1adeaa43
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-54801
Signed-off-by: Lei,RayX <rayx.lei@intel.com>
---
 AndroidManifest.xml                                     |  1 +
 .../android/car/dialer/telecom/InCallServiceImpl.java   | 17 +++++++++++++++++
 2 files changed, 18 insertions(+)

diff --git a/AndroidManifest.xml b/AndroidManifest.xml
index 571ab9b..59e3686 100644
--- a/AndroidManifest.xml
+++ b/AndroidManifest.xml
@@ -27,6 +27,7 @@
     <uses-permission android:name="android.permission.MODIFY_PHONE_STATE"/>
     <uses-permission android:name="android.permission.CALL_PHONE"/>
     <uses-permission android:name="android.permission.INTERACT_ACROSS_USERS"/>
+    <uses-permission android:name="android.permission.EXPAND_STATUS_BAR" />
 
     <!-- The Dialer needs to be directBootAware so that it can reflect the correct call state
          when the system boots up. -->
diff --git a/src/com/android/car/dialer/telecom/InCallServiceImpl.java b/src/com/android/car/dialer/telecom/InCallServiceImpl.java
index 4868896..1fde2bb 100644
--- a/src/com/android/car/dialer/telecom/InCallServiceImpl.java
+++ b/src/com/android/car/dialer/telecom/InCallServiceImpl.java
@@ -23,6 +23,9 @@ import android.telecom.CallAudioState;
 import android.telecom.InCallService;
 import android.telecom.TelecomManager;
 import android.util.Log;
+import android.os.ServiceManager;
+import android.os.RemoteException;
+import com.android.internal.statusbar.IStatusBarService;
 
 import java.util.concurrent.CopyOnWriteArrayList;
 
@@ -95,6 +98,20 @@ public class InCallServiceImpl extends InCallService {
                     Log.i(TAG, "Incoming/outgoing call: " + call);
                 }
 
+                if (state == Call.STATE_RINGING) {
+                    // Avoiding the fullscreen statusbar hide the incoming call.
+                    // Simply call collapsePanels to collapse the statusbar.
+                    IStatusBarService service = IStatusBarService
+                            .Stub.asInterface(ServiceManager.getService("statusbar"));
+                    if (service != null) {
+                        try {
+                            service.collapsePanels();
+                        } catch (RemoteException e) {
+                            Log.e(TAG, "Failed to collapsePanels for statusbar");
+                        }
+                    }
+                }
+
                 // TODO(b/25190782): here we should show heads-up notification for incoming call,
                 // however system notifications are disabled by System UI and we haven't implemented
                 // a way to show heads-up notifications in embedded mode.
-- 
1.9.1

